### PR TITLE
Base64.Decode: fixed latent bug for non-ASCII inputs

### PR DESCRIPTION
**Pull request by gfoidl**
Repro:
```c#
string base64      = "ìz/TpH7sqEkerqMweH1uSw==";    // first char is not ASCII
byte[] base64Bytes = Encoding.UTF8.GetBytes(base64);
Span<byte> data    = stackalloc byte[128];

OperationStatus status = Base64_.DecodeFromUtf8(base64Bytes, data, out int consumed, out int written);
```
Here the `ì` isn't recognized as being not part of the base64-alphabet, so wrong data is decoded. Ultimately the check https://github.com/dotnet/runtime/blob/f33d77841d936570a94af2a4c567da7c6d3f7045/src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Decoder.cs#L200 fails as the length doesn't match (base64-len is 24, base64Bytes-len is 25).
So status reports `InvalidData` correctly, but consumed (24) and written (17) are wrong, as they should be 0 each.

This latent bug got introduced by https://github.com/dotnet/runtime/pull/70654, and unfortunately there was a test-hole that didn't catch this.
https://github.com/dotnet/runtime/pull/70654#discussion_r895762249 talks about the shuffle mask -- here the highest bit of the mask-byte is set so after the shuffle the bits of that vector-element are set to `0`, thus https://github.com/dotnet/runtime/blob/f33d77841d936570a94af2a4c567da7c6d3f7045/src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Decoder.cs#L598-L601 doesn't detect the non-ASCII / non-base64 char.

By masking with `0xF` we clear the highest bit -- actually we mask by `0x2F` as that constant is used somewhere else and for shuffling (besides the highest bit) only the lower-nibble of the mask-byte is relevant.

### CODE REVIEW

File: src/libraries/System.Memory/tests/Base64/Base64DecoderUnitTests.cs [hunk 1]
```diff
@@ -272,6 +272,23 @@ public void BasicDecodingWithFinalBlockTrueKnownInputInvalid(string inputString,
             Assert.True(Base64TestHelper.VerifyDecodingCorrectness(expectedConsumed, decodedBytes.Length, source, decodedBytes));
         }
 
+        [Theory]
+        [InlineData("�z/T", 0, 0)]                                              // scalar code-path
```
Comment (gfoidl) [inline src/libraries/System.Memory/tests/Base64/Base64DecoderUnitTests.cs:276]: Ah, I need to update to literals here and the other places.

**Comment by gfoidl** [comment]
In CI there's lots of
```
{"$id":"1","innerException":null,"message":"VS403205: The organization is currently offline for migration. Once the migration has been completed the organization will come back online. Try accessing the organization at a later time. Host Id: 6fcc92e5-73a7-4f88-8d13-d9045b45fb27","typeName":"Microsoft.TeamFoundation.Framework.Server.HostShutdownException, Microsoft.TeamFoundation.Framework.Server","typeKey":"HostShutdownException","errorCode":0,"eventId":3000}
   at Microsoft.DotNet.Helix.AzureDevOps.AzureDevOpsTask.HandleFailedRequest(HttpRequestMessage req, HttpResponseMessage res) in /_/src/Microsoft.DotNet.Helix/Sdk/AzureDevOpsTask.cs:line 139
   at Microsoft.DotNet.Helix.AzureDevOps.AzureDevOpsTask.ParseResponseAsync(HttpRequestMessage req, HttpResponseMessage res) in /_/src/Microsoft.DotNet.Helix/Sdk/AzureDevOpsTask.cs:line 147
   at Microsoft.DotNet.Helix.AzureDevOps.StartAzurePipelinesTestRun.<>c__DisplayClass32_0.<<ExecuteCoreAsync>b__0>d.MoveNext() in /_/src/Microsoft.DotNet.Helix/Sdk/StartAzurePipelinesTestRun.cs:line 59
--- End of stack trace from previous location ---
```
When the org is online again, someone please re-trigger the build. Thanks in advance.

**Comment by kasperk81** [comment]
> When the org is online again

no clue. maintenance breaks should be publicly announced either at https://github.com/dotnet/announcements/issues or https://github.com/dotnet/core/issues

@MattGal

File: src/libraries/System.Memory/tests/Base64/Base64DecoderUnitTests.cs [hunk 2]
```diff
@@ -272,6 +272,23 @@ public void BasicDecodingWithFinalBlockTrueKnownInputInvalid(string inputString,
...
+        [InlineData("\u00ecz/TpH7sqEkerqMweH1uSw1a5ebaAF9xa8B0ze1wet4epo==", 0, 0)]  // Vector256 / AVX code-path
+        [InlineData("z/TpH7sqEkerqMweH1uSw1a5ebaAF9xa8B0ze1wet4epo\u00ec==", 44, 33)]
+        public void BasicDecodingNonAsciiInputInvalid(string inputString, int expectedConsumed, int expectedWritten)
+        {
+            Span<byte> source = Encoding.UTF8.GetBytes(inputString);
+            Span<byte> decodedBytes = new byte[Base64.GetMaxDecodedFromUtf8Length(source.Length)];
+
+            Assert.Equal(OperationStatus.InvalidData, Base64.DecodeFromUtf8(source, decodedBytes, out int consumed, out int decodedByteCount));
+            Assert.Equal(expectedConsumed, consumed);
+            Assert.Equal(expectedWritten, decodedByteCount); // expectedWritten == decodedBytes.Length
```
Comment (adamsitnik) [inline src/libraries/System.Memory/tests/Base64/Base64DecoderUnitTests.cs:289]: the comment seems to be invalid, as `expectedWritten` is not always equal `decodedBytes.Length`

```suggestion
            Assert.Equal(expectedWritten, decodedByteCount);
```

**Review by adamsitnik — APPROVED** [review]
LGMT, thank you for the fix @gfoidl !

**Comment by MattGal** [comment]
> > When the org is online again
> 
> no clue. maintenance breaks should be publicly announced either at https://github.com/dotnet/announcements/issues or https://github.com/dotnet/core/issues
> 
> @MattGal

Things are back to normal now, but yes I agree this was less than ideal.  The expectation from previous migrations was that this should have taken a few hours, but once it started it had to be completed.  It was expected that an Azure Devops Banner and partners DL email was enough, but clearly it caught people by surprise and we'll take these learnings for the next time a scheduled downtime happens.

**Comment by stephentoub** [comment]
> It was expected that an Azure Devops Banner and partners DL email was enough

There was no indication anywhere in dotnet/runtime, at least not that I saw.

**Pull request merged**